[front] - chore(mentions): improve mention suggestions

JulesBelveze opened the pull request:
## Description

This PR improves the mention suggestion dropdown to better prioritize relevant users and agents in conversations. 

The main changes include: prioritizing conversation participants (both users and agents) at the top of the dropdown with a cap of 5 total participants, implementing a 30/70 split between users and agents in suggestion results to ensure user visibility, filtering out the current user from suggestions to prevent self-mentions, promoting the preferred agent (last used) to the first position, and adding visual improvements like a "User" chip to distinguish user mentions and auto-scrolling of selected items.

**References:**
- https://github.com/dust-tt/tasks/issues/5328

## Risk

The logic for mixing and prioritizing suggestions has been significantly refactored in both the frontend (`MentionDropdown.tsx`) and backend (`mention_suggestions.ts`). There's a risk of edge cases in the ordering logic, particularly when handling conversations with few or no participants.

## Tests

- [x] Locally
- [x] front-edge 

## Deploy Plan

Deploy front

JulesBelveze committed:
[api/assistant/conversation] - feature: update mention suggestion logic to mix agents and users

 - Modify suggestion logic to allow separate agent or user suggestions based on query parameters
 - Implement a 30%/70% split algorithm for combining agent and user suggestions when both are requested
 - Ensure fallback to agent suggestions if no user suggestions are available
 - Fix an issue where `conversation` was not updated in the mentions suggestion API endpoint

JulesBelveze committed:
[front] - feature: enhance UI for user mentions in editor input bar

 - Update styling for the mention suggestions to improve readability and aesthetics
 - Introduce a "User" chip next to user mentions for clearer identification

JulesBelveze committed:
[api/assistant] - feature: prioritize conversation participants in mention suggestions

 - Reorder mention suggestions to favor current conversation participants
 - Fetch and utilize the conversation context to enhance mention suggestions

[api/analytics] - refactor: improve clarity of programmatic cost calculation

 - Adjust multi-line conditional expression for cutoffTime clarity in cost calculations

[api] - fix: correct API endpoint for conversation mention suggestions

 - Standardize API endpoint handler behavior for fetching conversation data

[developers/credits-usage] - refactor: enhance readability of credits usage page code

 - Simplify progress bar component color setting with a conditional class
 - Utilize consistent formatting for billing cycle date representations
 - Adjust renewalDate and cycleEnd calculation for better consistency

JulesBelveze committed:
[api/assistant/conversation] - fix: ensure proper conversation data structure for mention suggestions

 - Refactor mention suggestion logic to use `toJSON` method for conversation data
 - Adjust the alignment of the participant user IDs mapping for readability

JulesBelveze committed:
[assistant/conversation] - feature: improve mix of users and agents in mention suggestions

 - Ensure a user always appears first in mention suggestions if available
 - Shuffle remaining users and agents to provide a diverse set of suggestions while maintaining the designated proportions

JulesBelveze committed:
[assistant/conversation] - feature: improve mention suggestions in conversations

 - Normalize the user query to lowercase for consistent suggestion filtering
 - Extend user suggestion logic to include participants not previously suggested
 - Reorder suggestions by participants' IDs for relevance
 - Make agents fallback when no user mention suggestions are available

JulesBelveze committed:
[front/components/editor/input_bar] - feature: prioritize preferred agent in mention dropdown

 - Suggestions in the mention dropdown are now reordered to prioritize a preferred agent if one exists
 - Participants of a conversation are prioritized in the suggestions, with a cap of 5 mentions
 - Suggestions now exclude the current user to prevent self-mentions
 - The preferred agent is now represented by their ID in the ordered suggestions, ensuring their prominence

[front/lib/api/assistant/conversation] - refactor: adjust mention suggestions logic for consistency and efficiency

 - Simplified the process of filtering and reordering user and agent suggestions by using sets for participant IDs
 - Implemented a shuffle mechanism to mix users and agents while preserving the count proportions and participant priority
 - Improved the suggestion mechanism by considering conversation context and avoiding duplication of participants

[front/components/assistant/conversation/input_bar] - feature: pass preferred agent ID to input bar container

 - Included preferred agent's ID in the input bar container props to enable smarter mention suggestions handling

[front/components/editor/input_bar] - refactor: improve mention suggestion plugin creation with preferred agent

 - Extended the mention suggestion plugin to accept a preferred agent id as an optional parameter for better suggestion targeting

JulesBelveze committed:
[editor] - feature: prioritize mention suggestions based on conversation participants

 - Promote conversation participants in the mention dropdown by moving them to the top of the list
 - Sort the preferred agent (last used) to the very first position if it is present within the suggestion list
 - Remove obsolete code that reorders mentions based on conversation participants from the suggestions API
 - Streamline the process of mixing users and agents suggestions following a 30/70 proportion for a better user experience

JulesBelveze committed:
[front/components/assistant] - fix: correct preferredAgentId reference in InputBarContainer

 - Update the property used for preferredAgentId to match the current agent configuration naming

[front/lib/api/assistant] - refactor: clean up mention suggestions imports and logic

 - Move lodash shuffle import to the top for consistency
 - Refine filteredAgents assignment to be constant for better readability and immutability practices

JulesBelveze committed:
[front] - refactor: improve MentionDropdown participant handling

 - Redesign the process for promoting conversation participants in the suggestion list to prioritize ones matching the query
 - Extend the logic to include participants that may not be returned by the global suggestion endpoint

JulesBelveze committed:
[editor] - fix: prevent mentioning self in conversation participants

 - Filter out the logged-in user from the participants' list in mention suggestions
 - Adjust the mentions ordering logic to account for the user's exclusion
 - Enhance the user experience by avoiding self-mentions in conversations

JulesBelveze committed:
[front/components/editor/input_bar] - refactor: update mention dropdown participant logic

 - Limit the number of users and agents shown in the mention dropdown
 - Separate user and agent participants for individual filtering and capping
 - Introduce constants to define the max top participants, users, and remaining slots for agents

JulesBelveze committed:
[front/components/assistant] - fix: update agent ID reference in input bar container

- Change from `configurationId` to `id` property to presumably align with updated data model

[front/components/editor] - feature: implement scrolling for selected item in mention dropdown

- Add a `useEffect` hook to scroll the selected item into view when it changes
- Ensure selectedItemRef is assigned only to the selected item for correct scroll behavior

[front/lib/api] - refactor: improve mention suggestion handling and constants

- Replace lodash's shuffle with a custom shuffle function to remove dependency
- Introduce constants for ratio and minimum count of user mentions for better configurability
- Remove outdated code and clean up usage to better comply with current user models

[front] - style: clean up redundant code and improve readability

- Simplify the setting of `selectedIndex` by removing unnecessary parentheses
- Remove a filter condition redundancy by using optional chaining for user ID comparison

JulesBelveze committed:
fix: lint/format

JulesBelveze committed:
[front/assistant] - feature: implement conditional user mention functionality in editor components

 - Modify the `InputBarContainer` component to respect the `userMentionsEnabled` flag when setting `preferredAgentId`
 - Update the `MentionDropdown` component to filter user mentions based on the `userMentionsEnabled` flag
 - Adjust the suggestion prioritization logic in `MentionDropdown` to cater to the new user mentions condition
 - Pass `userMentionsEnabled` prop through various editor components to conditionally enable user mentions
 - Ensure `createMentionSuggestion` and `useCustomEditor` hooks receive and honor the `userMentionsEnabled` flag within editor extensions configuration

ElPicador commented on front/components/editor/input_bar/MentionDropdown.tsx:
```diff
@@ -19,165 +20,308 @@ import type {
+      workspaceId: owner.sId,
+      conversationId,
+      query,
+      select: { agents: true, users: userMentionsEnabled ?? false },
```
you can keep true, there is a check backend side in `front/pages/api/w/[wId]/assistant/conversations/[cId]/mentions/suggestions.ts`

ElPicador reviewed (approved):
I would move what is in the frontend to the backend, as per the other comment

JulesBelveze committed:
[assistant/conversation] - refactor: streamline mention suggestion logic

 - Removed redundant userMentionsEnabled flag to simplify mention suggestion configuration
 - Consolidated participant prioritization and preferred agent handling in the backend

[editor/input_bar] - refactor: remove user-specific suggestion filtering

 - Simplified MentionDropdown component by outsourcing suggestion filtering to the backend API
 - Enhanced mention suggestion logic to prioritize conversation participants and preferred agents

[api/assistant/conversation] - feature: implement backend logic for conversation participant prioritization

 - Added functionality to prioritize conversation participants in mention suggestions
 - Enabled automatic promotion of the preferred agent to the top of the suggestions list

[lib/api/assistant] - feature: extend mention suggestion API to include participant prioritization

 - Updated suggestion API to incorporate logic for conversation participants and preferred agents
 - Ensured compatibility with existing frontend components and API consumers
